toolchain/parse/node_kind.cpp

```diff
@@ -37,30 +36,9 @@ CARBON_DEFINE_ENUM_CLASS_NAMES(NodeKind) = {
 #include "toolchain/parse/node_kind.def"
 };
 
-auto NodeKind::CheckMatchesTokenKind(Lex::TokenKind token_kind, bool has_error)
-    -> void {
-  static constexpr Lex::TokenKind TokenIfValid[] = {
-#define CARBON_IF_VALID(LexTokenKind) LexTokenKind
-#define CARBON_PARSE_NODE_KIND_WITH_TOKEN(Name, LexTokenKind) \
-  Lex::TokenKind::LexTokenKind,
+// Check that each typed node defines a `token` member.
+#define CARBON_PARSE_NODE_KIND(Name) static_assert(&Name::token != nullptr);
```

Comment:
Does this need the nullptr comparison? FWIW I think we often don't do `!= nullptr` in conditions.

```suggestion
#define CARBON_PARSE_NODE_KIND(Name) static_assert(&Name::token);
```

Reply:
To me, forming a pointer-to-member in a static assert condition looked wrong, because it clearly can't ever be null. Replaced with a requires-expression, which hopefully better captures the intent that we're checking a requirement on the type here.